File: nQueen.cpp
//time complexity:O(n!)
//space complexity:O(n)
//executed on leetcode: yes
//approach:using backtracking
//any issues faced? yes

class Solution {
public:
    vector<vector<string>>res;
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<int>>board(n,vector<int>(n));
        set<int>diagonal1;
        set<int>diagonal2;
        set<int>vertical;
        dfs(board,0,diagonal1,diagonal2,vertical);
        return res;
    }
    void dfs(vector<vector<int>>& board, int i, set<int>&diagonal1, set<int>diagonal2, set<int>&vertical)
    {
        if(i==board.size())
        {
            add(board);
            return;
        }
        for(int j=0; j<board.size(); j++)
        {
            if(diagonal1.find(i+j)==diagonal1.end() && diagonal2.find(j-i)==diagonal2.end() && vertical.find(j)==vertical.end())
            {
                board[i][j]=1;
                diagonal1.insert(i+j);
                diagonal2.insert(j-i);
                vertical.insert(j);
                dfs(board,i+1,diagonal1,diagonal2,vertical);
                board[i][j]=0;
                diagonal1.erase(i+j);
                diagonal2.erase(j-i);
                vertical.erase(j);
            }
        }
        
    }
    void add(vector<vector<int>>&board)
    {
        vector<string>curr;
        for(int i=0; i<board.size(); i++)
        {
            string temp="";
            for(int j=0; j<board.size(); j++)
            {
                if(board[i][j]==0)
                    temp+=".";
                else
                    temp+="Q";
            }
            curr.push_back(temp);
        }
        res.push_back(curr);
    }
};

File: n_queen.py
# Time Complexity: O(n!)
# Space COmplexity: O(n^2)
# Ran on Leetcode: Yes

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        self.res = []
        board = [["." for i in range(n)] for j in range(n)]
        
        self.col = [False] * n
        self.dia_l = [False] * (n * 2)
        self.dia_r = [False] * (n * 2)
        
        if n == 0:
            return []
        self.solver(board, 0 ,0, n)
        
        return self.res
    
    def solver(self, board, i, j, n):
        if i == n:
            self.res.append(["".join(row) for row in board])
            return
        
        while j < n:
            
            if self.valid(board, i, j):
                
                board[i][j] = 'Q'
                
                self.col[j] = True
                self.dia_r[i + j] = True
                self.dia_l[i - j] = True

                self.solver(board, i + 1, 0, n)

                board[i][j] = '.'
                self.col[j] = False
                self.dia_r[i + j] = False
                self.dia_l[i - j] = False
                
            j += 1
    
    def valid(self, board, i, j):
        if self.col[j] or self.dia_r[i + j] or self.dia_l[i - j]:
            return False
        return True
        
        